### ISSUE TITLE
win: arithmetic overflow in uv__hrtime()?

### ISSUE THREAD

**Issue by bnoordhuis** [libuv/libuv#2843]
https://github.com/libuv/libuv/blob/36549815eed297c74448ff580b5c84713245ec5d/src/win/util.c#L501-L509

The calculation looks suspect to me, it overflows when the clock frequency and scale are high.

It also uses the performance counter as-is (which might be big) without subtracting a start time.

With `interval=1e7,` (1 increment per 100 ns) and `scale=1e9`, overflows happens when the counter exceeds 922 because `922 * 1e7 * 1e9 ~= 2**63` - except it's not exactly `2**63` because it's floating-point arithmetic.

That might or might not explain #1633. It uses `scale=1e3`. `double` is precise up to `2**53`, after that it becomes imprecise.

Maybe it's producing Infinity or NaN and getting truncated to zero by the cast to `uint64_t`?

I think the correct logic would look something like this:
```c
static LARGE_INTEGER hrtime_base;
static LARGE_INTEGER hrtime_freq;

// at startup
QueryPerformanceCounter(&hrtime_base);
QueryPerformanceFrequency(&hrtime_freq);

// in uv__hrtime()
QueryPerformanceCounter(&now);
now.QuadPart -= hrtime_base.QuadPart;  // note this part!

__int128 t = (__int128) now.QuadPart;
t *= (__int128) scale;
t /= (__int128) hrtime_freq.QuadPart;
return (uint64_t) t;
```

cc @libuv/windows @addaleax

**Comment by bzoz** [comment]
Yeah, I saw that too and my first thought was that if the counter starts big we would run into problems.  

I guess we should subtract the initial value and move from double anyway. There is no `__int128_t` in VS, but there are 128-bit things like [_mul128](https://docs.microsoft.com/pl-pl/cpp/intrinsics/mul128?view=vs-2019) and [_udiv128](https://docs.microsoft.com/en-us/cpp/intrinsics/udiv128?view=vs-2019), but that last one is only available since VS2019.

**Comment by erw7** [comment]
If `interval=1e7`, that's 1 increment per 116 days (`1e7 / (60 * 60 * 24) ~= 116`). At a `frequency = 1e7 Hz` (1e7 count per second, 1 increment per 100 ns), the `interval = 1.0 / 1e7 = 1e-7  sec = 100ns`. Thus, even if `scale = 1e9`, the expression is `counter.QuadPart * 1e-7 * 1e9 = counter.QuadPart * 1e2`. Even when `scale = 1e9`, it take about 585 (`2**64 / (365 * 24 * 60 * 60 * 1e9)`) years for the `uin64_6` type to overflow.

I'm getting less and less confident in my calculations as I write, but we shouldn't have to worry about overflow here anyway.

**Comment by bzoz** [comment]
https://github.com/libuv/libuv/issues/1633#issuecomment-628147954 - the initial values are not that big, `hrtime_freq` has the same value as on my box that does not reproduce #1633.

**Comment by bnoordhuis** [comment]
@bzoz I agree, but I think the calculation is indeed incorrect. That numbers the reporter posted work out to `272692642196 * 1e7 * 1e3`, or about 2**71. Definitely doesn't fit in an `uint64_t`.

The counterargument of course is that the numbers you posted are even worse and your machine doesn't exhibit the problem. :-)

Your observation about 128 bits math being complicated is on point. I couldn't even get mingw to recognize `__int128` or `int __attribute__((__mode__(TI)))`, although it supposedly supports it. I'll try to poke at it more tomorrow.

**Comment by erw7** [comment]
1e7 is the frequency. As is clear from the code below, `hrtime_interval_` is 1e-7(= 1.0 / 1e7). Again, there is absolutely no way that uint64_t will overflow here.

https://github.com/libuv/libuv/blob/36549815eed297c74448ff580b5c84713245ec5d/src/win/util.c#L86-L88

**Comment by bnoordhuis** [comment]
Oh, you're absolutely right! Yes, in that case it's on the order of 2**45 with `scale=1e9`, and only 2**25 bits with `scale=1e3`. Okay, there goes that hypothesis out the window.
